src/edu/washington/escience/myria/operator/agg/StreamingAggregate.java

```diff
@@ -0,0 +1,199 @@
+
+  /** Group fields. **/
+  private final int[] gFields;
+  /** An array [0, 1, .., gFields.length-1] used for comparing tuples. */
+  private final int[] gRange;
+  /** Factories to make the Aggregators. **/
+  private final AggregatorFactory[] factories;
+  /** The actual Aggregators. **/
+  private Aggregator[] aggregators;
+  /** The state of the aggregators. */
+  private Object[] aggregatorStates;
+
+  /** Buffer for holding intermediate results. */
+  private transient TupleBatchBuffer resultBuffer;
+
+  /**
+   * Groups the input tuples according to the specified grouping fields, then produces the specified aggregates.
+   *
+   * @param child The Operator that is feeding us tuples.
+   * @param gfields The columns over which we are grouping the result.
+   * @param factories The factories that will produce the {@link Aggregator}s for each group.
+   */
+  public StreamingAggregate(@Nullable final Operator child, @Nonnull final int[] gfields,
+      @Nonnull final AggregatorFactory... factories) {
+    super(child);
+    gFields = Objects.requireNonNull(gfields, "gfields");
+    this.factories = Objects.requireNonNull(factories, "factories");
+    Preconditions.checkArgument(gfields.length > 0, " must have at least one group by field");
+    Preconditions.checkArgument(factories.length > 0, "to use StreamingAggregate, must specify some aggregates");
+    gRange = new int[gfields.length];
+    for (int i = 0; i < gRange.length; ++i) {
+      gRange[i] = i;
+    }
+  }
+
+  /**
+   * Returns the next tuple batch containing the result of this aggregate. Grouping field(s) followed by aggregate
+   * field(s).
+   *
+   * @throws DbException if any error occurs.
+   * @return result tuple batch
+   */
+  @Override
+  protected TupleBatch fetchNextReady() throws DbException {
+    if (child.eos()) {
+      return resultBuffer.popAny();
+    }
+    if (tb == null) {
+      tb = child.nextReady();
+      row = 0;
+    }
+    while (tb != null) {
+      while (row < tb.numTuples()) {
+        if (curGroupKey == null) {
+          /* First time accessing this tb, no aggregation performed previously. */
+          // store current group key as a tuple
+          curGroupKey = new Tuple(groupSchema);
+          for (int gKey = 0; gKey < gFields.length; ++gKey) {
+            TupleUtils.copyValue(tb, gFields[gKey], row, curGroupKey, gKey);
+          }
+        } else if (!TupleUtils.tupleEquals(tb, gFields, row, curGroupKey, gRange, 0)) {
+          /* Different grouping key than current one, flush current agg result to result buffer. */
+          addToResult();
+          // store current group key as a tuple
+          for (int gKey = 0; gKey < gFields.length; ++gKey) {
+            TupleUtils.copyValue(tb, gFields[gKey], row, curGroupKey, gKey);
+          }
+          aggregatorStates = AggUtils.allocateAggStates(aggregators);
+          if (resultBuffer.hasFilledTB()) {
+            return resultBuffer.popFilled();
+          }
```

Comment:
`fetchNextReady()` is changed such that as soon as the aggregator has filled a `TupleBatch`, it will return that `TupleBatch` and pause aggregation at that point. The next call to `fetchNextReady()` resumes aggregation until it fills another `TupleBatch`. Unfilled `TupleBatch` is expected only when `child.eos()` is reached.